qiskit_nature/deprecation.py

```diff
@@ -0,0 +1,242 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Contains the Deprecation message methods."""
+
+import warnings
+import functools
+from typing import NamedTuple, Optional, Callable, Dict, Set
+from enum import Enum
+
+
+class DeprecatedType(Enum):
+    """ " Deprecation Types"""
+
+    PACKAGE = "package"
+    ENUM = "enum"
+    CLASS = "class"
+    METHOD = "method"
+    FUNCTION = "function"
+    ARGUMENT = "argument"
+    PROPERTY = "property"
+
+
+class _DeprecatedObject(NamedTuple):
+    version: str
+    old_type: DeprecatedType
+    old_name: str
+    new_type: DeprecatedType
+    new_name: str
+    additional_msg: str
+
+
+_DEPRECATED_OBJECTS: Set[_DeprecatedObject] = set()
+
+
+def warn_deprecated(
+    version: str,
+    old_type: DeprecatedType,
+    old_name: str,
+    new_type: Optional[DeprecatedType] = None,
+    new_name: Optional[str] = None,
+    additional_msg: Optional[str] = None,
+    stack_level: int = 2,
+) -> None:
+    """Emits deprecation warning the first time only
+    Args:
+        version: Version to be used
+        old_type: Old type to be used
+        old_name: Old name to be used
+        new_type: New type to be used
+        new_name: New name to be used
+        additional_msg: any additional message
+        stack_level: stack level
+    """
+    # skip if it was already added
+    obj = _DeprecatedObject(version, old_type, old_name, new_type, new_name, additional_msg)
+    if obj in _DEPRECATED_OBJECTS:
+        return
+
+    _DEPRECATED_OBJECTS.add(obj)
+
+    msg = (
+        f"The {old_name} {old_type.value} is deprecated as of version {version} "
+        "and will be removed no sooner than 3 months after the release"
```

Comment:
Maybe just a general question, but this seems like a good place to ask: where do these 3 months come from? Is this because of the approximate 3-month release cycle? Wouldn't it make more sense to tie the deprecation to the next release (e.g. `0.{n+1}.0` if deprecated with release of `0.{n}.0`). Given our dev cycles this will roughly coincide with 3 months but it will be *much* easier to keep track off and will also not require users to know when version `0.{n}.0` actually came out in order to know when this will be removed..

Reply:
I believe it is a standard set initially by Terra. I agree that tying the schedule to some release after seems better than an arbitrary timeframe. Maybe @woodsp-ibm would have more to say about it.

Reply:
I agree with @mrossinek. Setting deprecation based on releases is reasonable and easier for people to follow.

Reply:
https://qiskit.org/documentation/contributing_to_qiskit.html#deprecation-policy

Reply:
Given the release when it is deprecated in there, which can be looked up and its date found, computing at least 3 months from that can be done. Ideally users should transition to the new way asap but the idea here is that get some time to do that. Saying in some future release by version people may have no idea when that is coming out - also plans change. In the end if you get deprecated messages you should be changing asap, if you procrastinate then things will just break when its removed. This clients see sometimes if they have some old code that worked and then go back to it after a few months during which time code it uses has since been both deprecated and removed after the time was up.

Reply:
Given that this is a documented Qiskit policy, I believe we need to comply with it then.

Reply:
> which can be looked up and its date found
I agree, but I think that this is unnecessary manual.

> may have no idea when that is coming out - also plans change
True, but I don't think this really affects what I proposed above

> This clients see sometimes if they have some old code that worked...
I actually think that having the version number as part of the deprecation warning would actually help with this situation as it makes clearer which versions to look for in the Changelog.

> https://qiskit.org/documentation/contributing_to_qiskit.html#deprecation-policy
Thanks! I see, if this is done the same everywhere as is, we should probably stick to it. I just wanted to bring it up since I personally would find it much more sensible, with the removal pinned to a version number. We could even include that link in the deprecation message..

Reply:
Its consistent with Qiskit. Maybe having a future date in there would be better - e.g. we know the date of Python 3.6 EOL so its easy to deal with. By plans changing I meant its harder to predict the version we will be 3 months out to put in the message - if we decide to bring the next one release sooner then it kinda messes things up in that regard. But we would know just ahead of release when 3 months out was to set a date for all deprecation messages added that release to use. But unless Qiskit overall changes I think though we should stick to the current way/message and be consistent.

Reply:
Yes let's leave it as is, since it appears to be Qiskit standard.

Just as a final comment to clarify: my suggestion was to make the deprecation such that we remove it at the next feature release **regardless** of when that is. Usually this fits well with 3 months but if it happens after 1 month then the user just needs to hurry up. Since this is really not in-line with Qiskit's deprecation policy I retract this proposal.